### PR TITLE
Containerise build process

### PR DESCRIPTION
**Pull request by cyqsimon**
## Describe your changes

Containerise the build process. Motivation detailed [here](https://github.com/pouriyajamshidi/tcping/pull/245#discussion_r1791118412), posted here verbatim:

> Currently, the build process itself is not performed in a containerised environment, which goes against the key spirits of containerisation - platform-independence and reproducibility. In short, the build process in its current state is dependent on a working toolchain on the host, which alongside its other environmental factors is largely opaque to the end user (or even to the packager themselves). Setting up the build environment in a containerised environment improves the observability, reliability, and trustworthiness of the build artefact(s).
>
> As of how a multi-stage Dockerfile fits into this equation, this is just so that we don't ship a massive image that unnecessarily contains the toolchain to the user. The build happens in `imageA`; then we copy the relevant artefacts into `imageB` and ship it. It's very well supported at this point and not at all complicated to set up.

## Issue ticket number and link

N/A

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added tests.
- [x] I have run `make check` and there are no failures.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

I guess this counts as a bug fix?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,7 +1,26 @@
+# Build stage
+##################################################
+FROM docker.io/golang:1.23-alpine3.20 AS build
```
Comment (cyqsimon) [inline Dockerfile:3]: If you don't want to have to manually bump this from time to time, [there are other tags available](https://hub.docker.com/_/golang) in varying degrees of brevity.

It's a tradeoff between convenience and reproducibility. Your call.

Comment (pouriyajamshidi) [reply Dockerfile:3]: what is the image size difference between this and the current approach?

Comment (cyqsimon) [reply Dockerfile:3]: > what is the image size difference between this and the current approach?

Basically identical, which is to be expected.

```
❯ podman images -n --format "{{.Repository}}:{{.Tag}} {{.Size}}" | rg tcping
localhost/tcping-new:latest 9.61 MB
ghcr.io/pouriyajamshidi/tcping:latest 9.62 MB
```

**Review by pouriyajamshidi — APPROVED** [review]
Thank you @cyqsimon

**Pull request merged**